tests/msc2716_test.go

```diff
@@ -0,0 +1,643 @@
+// +build msc2716
+
+// This file contains tests for incrementally importing history to an existing room,
+// a currently experimental feature defined by MSC2716, which you can read here:
+// https://github.com/matrix-org/matrix-doc/pull/2716
+
+package tests
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"strconv"
+	"testing"
+	"time"
+
+	"github.com/matrix-org/complement/internal/b"
+	"github.com/matrix-org/complement/internal/client"
+	"github.com/matrix-org/complement/internal/match"
+	"github.com/matrix-org/complement/internal/must"
+	"github.com/tidwall/gjson"
+)
+
+type event struct {
+	Type           string
+	Sender         string
+	OriginServerTS uint64
+	StateKey       *string
+	PrevEvents     []string
+	Content        map[string]interface{}
+}
+
+// This is configurable because it can be nice to change it to `time.Second` while
+// checking out the test result in a Synapse instance
+const TimeBetweenMessages = time.Millisecond
+
+var (
+	MSC2716_INSERTION = "org.matrix.msc2716.insertion"
+	MSC2716_MARKER    = "org.matrix.msc2716.marker"
+
+	MSC2716_HISTORICAL                   = "org.matrix.msc2716.historical"
+	MSC2716_NEXT_CHUNK_ID                = "org.matrix.msc2716.next_chunk_id"
+	MSC2716_CHUNK_ID                     = "org.matrix.msc2716.chunk_id"
+	MSC2716_MARKER_INSERTION             = "org.matrix.msc2716.marker.insertion"
+	MSC2716_MARKER_INSERTION_PREV_EVENTS = "org.matrix.msc2716.marker.insertion_prev_events"
+)
+
+// Test that the message events we insert between A and B come back in the correct order from /messages
+func TestBackfillingHistory(t *testing.T) {
+	deployment := Deploy(t, b.BlueprintHSWithApplicationService)
+	defer deployment.Destroy(t)
+
+	// Create the application service bridge user that is able to backfill messages
+	asUserID := "@the-bridge-user:hs1"
+	as := deployment.Client(t, "hs1", asUserID)
+
+	// Create the normal user which will send messages in the room
+	userID := "@alice:hs1"
+	alice := deployment.Client(t, "hs1", userID)
+
+	// Create the federated user which will fetch the messages from a remote homeserver
+	remoteUserID := "@charlie:hs2"
+	remoteCharlie := deployment.Client(t, "hs2", remoteUserID)
+
+	virtualUserLocalpart := "maria"
+	virtualUserID := fmt.Sprintf("@%s:hs1", virtualUserLocalpart)
+
+	t.Run("parallel", func(t *testing.T) {
+		t.Run("Backfilled historical events resolve with proper state in correct order", func(t *testing.T) {
```

Comment:
I don't have a clear mental model of what state you want the room to end up in. It would be infinitely better to add a small ASCII diagram comment at this line to say something like:
```go
// Final timeline output: ( [n] = historical chunk )
// A,B [c, d, e] [f, g, h], I J
//      chunk 2   chunk 1
```
which is what I _think_ is the resulting output?